Fix connection stack in stock worker

kampfschlaefer opened the pull request:
When running the stock worker itself, the `_connection_stack` is empty. Sadly this only shows when the `rqworker` runs in its own python process. The existing tests in test_worker.py and test_cli.py don't do that.

This branch adds tests that run the stock worker in its own subprocess, thus showing the problem. And it extends the cli to push (and pop) the redis connection when running the worker.

Also I changed the existing `access_self` function to also assert on `get_current_connection`. And this extends the existing worker test to check the failed queue to see that no job failed.

This fixes #479.

kampfschlaefer commented:
okay, I have to figure out how to start the rqworker without it being installed. and without knowing how the python binary is called. in tox it worked...

selwin commented on tests/fixtures.py:
```diff
@@ -55,9 +55,15 @@ def create_file_after_timeout(path, timeout):
     assert get_current_job() is not None
 
 
+def schedule_access_self():
```
Since this function is only used in one place, could you please move this function definition to within `test_run_scheduled_access_self`?

selwin commented:
@kampfschlaefer I just read the comments of #479 again. The main issue is that the worker runs inside its own process which is unable to call `get_current_connection`, correct?

The current patch will only fix the issue for people running the stock `rqworker` command - and I know some people have their own worker scripts. I think it would be better to call `push_connection()` in `worker.prepare_job_execution()`. This should fix this bug for everyone. I haven't tried this in code though, do you think this would work?

kampfschlaefer commented:
> The current patch will only fix the issue for people running the stock rqworker command - and I know some people have their own worker scripts. I think it would be better to call push_connection() in worker.prepare_job_execution(). This should fix this bug for everyone. I haven't tried this in code though, do you think this would work?

I will try this the next days and update this PR.

Would adding this `push_connection` break anything for people that already do that in their own workers? Or would it just be one more item on the stack?

kampfschlaefer commented:
Hi,

I did the wanted changes. Also I pulled in master and did some fixes for flake8 to stop complaining (at least the flake8 as run by tox, my editor still has lots of red).

This PR should be ready for merging. Or you just merge the stuff before my reverse-merge. Or if wanted I can rebase it all against current master…

kampfschlaefer commented:
funny how the slow execution of the pypy tests marks the whole PR bad :-(

selwin commented:
@kampfschlaefer I re-ran the test on Travis and everything now passes. Could you please move `TestWorkerSubprocess` from `test_cli` to `test_worker` though? Thanks!

kampfschlaefer commented:
Ping. I rebased and squashed this PR into three commits. (By accident, got asked to rebase a completely different PR in another project but didn't read the mail correctly. :wink: )

Maybe the the cleaner history now is more suited for merging? (Apart from the pypy tests failing again…)

selwin commented:
Sorry, I totally forgot about this PR @kampfschlaefer . Could you please fix the failing pypy test and resolve the conflicts. Again, sorry for forgetting about this, I've been very busy with work lately.

kampfschlaefer commented:
I would love to fix the failing test. Unfortunately I can not reproduce the failure locally. I will rebase this PR against current master, maybe travis will not have a problem then.

kampfschlaefer commented:
Finally I got travis to be green on this PR. I have the feeling that the timeouts are highly volatile…

ping @selwin

selwin commented:
@kampfschlaefer I'm really sorry that I missed this PR. Due to recently merged changes, this PR no longer merges cleanly. Could you please resolve the merge conflicts so I can pull this in?

nvie commented on rq/worker.py:
```diff
@@ -635,6 +637,8 @@ def perform_job(self, job, queue):
         else:
             self.log.warning('Result will never expire, clean up result key manually')
 
+        pop_connection()
```
This should be called in a `finally:` block, to make sure it's always called, even in case of an exception.

kampfschlaefer commented:
Aaand I am back to slow pypy marking the PR red…

@selwin @nvie please restart the travis run or just merge this PR despite being red.

selwin commented on rq/worker.py:
```diff
@@ -560,6 +560,8 @@ def prepare_job_execution(self, job):
         """
         timeout = (job.timeout or 180) + 60
 
+        push_connection(self.connection)
```
One more minor thing. Should we move this statement to `perform_job()`, right before `prepare_job_execution()` is called? Having `push_connection()` and `pop_connection()` in the same scope makes more sense.

nvie replied:
+1 to that. The push/pop calls should always be balanced.

selwin commented:
I restarted the build and they all pass :)

kampfschlaefer commented:
Moved the `push_connection` call into `perform_job()`. Now its on the same logical level as the `finally` that contains the `pop_connection`.

selwin commented:
Merged, thanks!

dan-blanchard commented:
Could we please get a release with this bugfix? We're still pinned to 0.4.2 because of this issue.

nvie commented:
@dan-blanchard I just released rq==0.7.0 to PyPI, which includes this bugfix.  Let me know if that works for you!

dan-blanchard commented:
thanks @nvie!
